autorest/azure/async.go

```diff
@@ -0,0 +1,165 @@
+	OperationFailed string = "Failed"
+	// OperationSucceeded says the underlying opertion succeeded.
+	OperationSucceeded string = "Succeeded"
+)
+
+// OperationError provides additional detail when an operation fails or is canceled.
+type OperationError struct {
+	// Code provides an invariant error code useful for troubleshooting, aggregration, and analysis.
+	Code string `json:"code,omitempty"`
+	// Message indicates what error occurred and what can be done to address the issue.
+	Message string `json:"message,omitempty"`
+}
+
+// Error implements the error interface returnin a string containing the code and message.
+func (oe OperationError) Error() string {
+	return fmt.Sprintf("Azure Operation Error -- code(%v), message(%v)", oe.Code, oe.Message)
+}
+
+// OperationResource defines a resource describing the state of a long-running operation.
+type OperationResource struct {
+	// Id is the identifier used in a GET for the underlying resource.
+	ID string `json:"id,omitempty"`
+	// Name matches the last segment of the Id field (typically system generated).
+	Name string `json:"name,omitempty"`
+	// Status provides the state of the operation. Non-terminal states vary by resource;
+	Status string `json:"status,omitempty"`
+	// Properties, on operation success, optionally contains per-service / per-resource values.
+	Properties map[string]interface{} `json:"properties,omitempty"`
+	// Error provides additional detail if the operation is canceled or failed.
+	Error OperationError `json:"error,omitempty"`
+	// StartTime optionally provides the time the operation started.
+	StartTime date.Time `json:"startTime,omitempty"`
+	// EndTime optionally provides the time the operation completed.
+	EndTime date.Time `json:"endTime,omitempty"`
+	// PercentComplete optionally provides the percent complete between 0 and 100.
+	PercentComplete float64 `json:"percentComplete,omitempty"`
+}
+
+// IsCanceled returns true if the operation was canceled; false otherwise.
+func (or OperationResource) IsCanceled() bool {
+	return or.Status == OperationCanceled
+}
+
+// HasFailed returns true if the operation has failed; false otherwise.
+func (or OperationResource) HasFailed() bool {
+	return or.Status == OperationFailed
+}
+
+// HasSucceeded returns true if the operation has succeeded; false otherwise.
+func (or OperationResource) HasSucceeded() bool {
+	return or.Status == OperationSucceeded
+}
+
+// HasTerminated returns true if the operation has terminated; false otherwise.
+func (or OperationResource) HasTerminated() bool {
+	return or.IsCanceled() || or.HasFailed() || or.HasSucceeded()
+}
+
+// GetError returns an error if the operation was canceled or failed and nil otherwise.
+func (or OperationResource) GetError() error {
+	switch {
+	case or.IsCanceled():
+		return fmt.Errorf("azure: Operation was canceled -- error(%s)", or.Error.Error())
+	case or.HasFailed():
+		return fmt.Errorf("azure: Operation failed -- error(%s)", or.Error.Error())
```

Comment:
`...%s", or.Error.Error()` is overkill.
`azure: Operation failed: %v", or.Error` should be just enough.